middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -460,123 +368,158 @@ and let_cont_inline env res k h body =
+and let_cont_exn_handler env res k body vars handler ~catch_id arity =
+  let exn_var, extra_params =
+    match vars with
+    | (v, _) :: rest -> v, rest
+    | [] ->
+      (* See comment on [translate_raise], above. *)
+      Misc.fatal_errorf
+        "Exception continuation %a should have at least one argument"
+        Continuation.print k
+  in
+  let env_body, mut_vars = Env.add_exn_handler env k arity in
   let handler =
-    (* wrap the exn handler with reads of the mutable variables *)
+    (* Wrap the exn handler with reads of the mutable variables *)
     List.fold_left2
-      (fun handler (v, _) (p, _) ->
-        C.letin p ~defining_expr:(C.var v) ~body:handler)
-      handle extra_vars extra_params
+      (fun handler (mut_var, _) (extra_param, _) ->
+        C.letin extra_param ~defining_expr:(C.var mut_var) ~body:handler)
+      handler mut_vars extra_params
   in
   let body, res = expr env_body res body in
   let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
-  let trywith = C.trywith ~dbg ~kind:(Delayed id) ~body ~exn_var ~handler () in
-  (* define and initialize the mutable cmm variables used by an exn extra
-     args *)
+  let trywith =
+    C.trywith ~dbg ~kind:(Delayed catch_id) ~body ~exn_var ~handler ()
+  in
+  (* Define and initialize the mutable Cmm variables for extra args *)
   let cmm =
     List.fold_left
-      (fun cmm (v, k) ->
+      (fun cmm (mut_var, (kind : K.t)) ->
         (* CR mshinwell: Fix [provenance] *)
-        let v = Backend_var.With_provenance.create ?provenance:None v in
-        C.letin_mut v (C.machtype_of_kind k) (default_of_kind ~dbg k) cmm)
-      trywith extra_vars
+        let mut_var =
+          Backend_var.With_provenance.create ?provenance:None mut_var
+        in
+        let dummy_value =
+          match kind with
+          | Value -> C.int ~dbg 1
+          | Naked_number Naked_float -> C.float ~dbg 0.
+          | Naked_number Naked_int64 when Target_system.is_32_bit ->
+            C.unsupported_32_bit ()
+          | Naked_number
+              (Naked_immediate | Naked_int32 | Naked_int64 | Naked_nativeint) ->
+            C.int ~dbg 0
+          | Region | Rec_info ->
+            Misc.fatal_errorf "No dummy value available for kind %a" K.print
+              kind
+        in
+        C.letin_mut mut_var (C.machtype_of_kind kind) dummy_value cmm)
+      trywith mut_vars
   in
   cmm, res
 
 and let_cont_rec env res conts body =
-  (* Flush the env before anything to avoid inlining something inside of a
-     recursive cont (aka a loop), as it would increase the number of times the
+  (* Flush the env now to avoid inlining something inside of a recursive
+     continuation (aka a loop), as it would increase the number of times the
      computation is performed (even if there is only one syntactic
      occurrence) *)
+  (* CR-someday mshinwell: As discussed, the tradeoff here is not clear, since
+     flushing might increase register pressure. *)
```

Comment:
I'm not sure to_cmm is the best place to take this tradeoff into account. Since flambda's job is to estimate benefits from transformations (mainly inlining), is there a particular reason why that analysis would be in to_cmm and not in simplify (which could move the let-bindings inside the loop if it wants to) ?